src/migrate/Migrator.js

```diff
@@ -46,10 +47,15 @@ const CONFIG_DEFAULT = Object.freeze({
 // the migration.
 export default class Migrator {
   constructor(knex) {
-    this.knex = knex;
-    this.config = getMergedConfig(knex.client.config.migrations);
-    this.generator = new MigrationGenerator(knex.client.config.migrations);
-
+    // Clone knex instance and remove post-processing that is unnecessary for internal queries from a cloned config
+    this.knex = isFunction(knex)
+      ? knex.withUserParams(knex.userParams)
+      : Object.assign({}, knex);
```

Comment:
This is great! I attempted to clone knex but unsuccessfully in some prototyping code - I would receive an error (forgot what it was) attempting to make queries with my cloned version. Good to see how it is done here.

Reply:
Admittedly I'm abusing .withUserParams function, but given that it includes all the necessary black magic for cloning to actually produce a fully usable knex instance, that's the best option we have now. Maybe we should eventually replace .clone function with what .withUserParams currently does and call it under-the-hood.